Topology aware volume scheduler and PV controller changes

msau42 opened the pull request:
**What this PR does / why we need it**:
Scheduler and PV controller changes to support volume topology aware scheduling, as specified in kubernetes/community#1168

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #54435

**Special notes for your reviewer**:
* I've split the PR into logical commits to make it easier to review
* The remaining TODOs I plan to address next release unless you think it needs to be done now

**Release note**:
```release-note
Adds alpha support for volume scheduling, which allows the scheduler to make PersistentVolume binding decisions while respecting the Pod's scheduling requirements.  Dynamic provisioning is not supported with this feature yet.

Action required for existing users of the LocalPersistentVolumes alpha feature:
* The VolumeScheduling feature gate also has to be enabled on kube-scheduler and kube-controller-manager.
* The NoVolumeNodeConflict predicate has been removed.  For non-default schedulers, update your scheduler policy.
* The CheckVolumeBinding predicate has to be enabled in non-default schedulers.
```

@kubernetes/sig-storage-pr-reviews @kubernetes/sig-scheduling-pr-reviews

msau42 commented:
/assign @jsafrane @bsalamat @saad-ali

BenTheElder commented:
/retest

msau42 commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -1424,6 +1453,13 @@ func NewVolumeNodePredicate(pvInfo PersistentVolumeInfo, pvcInfo PersistentVolum
 func (c *VolumeNodeChecker) predicate(pod *v1.Pod, meta algorithm.PredicateMetadata, nodeInfo *schedulercache.NodeInfo) (bool, []algorithm.PredicateFailureReason, error) {
+
+	// Don't run this predicate if volume topology scheduling is enabled
+	// TODO msau: how do I properly remove this predicate???
```
@bsalamat I want to remove/replace this predicate that I added in 1.7 as an alpha feature, but it seems like the scheduler has some backwards compatibility policy.  Could you advise on what I should do here?

bsalamat replied:
What is the specific issue you are seeing when you remove the predicate?

msau42 replied:
The compatibility test fails because the predicate existed in 1.7 but not anymore in 1.9.  It seems like we cannot ever remove predicates.

bsalamat replied:
If the compatibility tests use a scheduler policy config and they don't update it in their upgrade process to remove the deleted predicate, they tests may fail. We must ensure that the compatibility test policy config is updated during the upgrade process.

msau42 replied:
In the compatibility [test](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/algorithmprovider/defaults/compatibility_test.go#L318), it says that we cannot change the json or else we'll break backwards compatibility.  Are you saying that since the predicate I added in 1.7 is an alpha feature and does not have upgrade guarantees, then it is an allowed exception to this policy?

Or is there some sort of scheduler upgrade process that I can modify?

msau42 replied:
I couldn't remove the predicate from defaults.go, unless I want to remove the predicate from the 1.7 compatibility test config.  It seems like the compatibility test is testing against the registered default predicates.

msau42 commented on plugin/pkg/scheduler/factory/plugins.go:
```diff
@@ -43,6 +44,8 @@ type PluginFactoryArgs struct {
 	NodeInfo                       predicates.NodeInfo
 	PVInfo                         predicates.PersistentVolumeInfo
 	PVCInfo                        predicates.PersistentVolumeClaimInfo
+	StorageClassInfo               predicates.StorageClassInfo
```
I don't know why the "Info" interface is needed instead of just passing in the Lister.  It seems like it's just a direct passthrough to the Lister...  But it's what was done for PV, PVC.

msau42 commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -379,6 +400,17 @@ func (c *configFactory) onPvcAdd(obj interface{}) {
 	}
 }
 
+func (c *configFactory) onPvcUpdate(oldObj interface{}, obj interface{}) {
```
This is not right.  Need to also think about how this works when invalidating the predicates after assume.

msau42 replied:
If we always invalidate after assumeVolume, then no need to invalidate on pvc update, so this can be removed.

msau42 commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -489,6 +508,16 @@ func (c *VolumeZoneChecker) predicate(pod *v1.Pod, meta algorithm.PredicateMetad
 
 			pvName := pvc.Spec.VolumeName
 			if pvName == "" {
+				if utilfeature.DefaultFeatureGate.Enabled(features.VolumeTopologyBinding) {
```
Add unit test

msau42 replied:
done

msau42 commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -1408,74 +1437,73 @@ type VolumeNodeChecker struct {
+//
+// The predicate returns true if all bound PVCs have compatible PVs with the node, and if all unbound
+// PVCs can be matched with an available and node-compatible PV.
+func NewVolumeBindingPredicate(binder persistentvolume.TopologyAwareVolumeBinder) algorithm.FitPredicate {
```
add unit test

msau42 replied:
added in scheduler_test.go

msau42 commented:
/retest

msau42 commented on plugin/pkg/scheduler/scheduler.go:
```diff
@@ -269,9 +299,84 @@ func (sched *Scheduler) assume(assumed *v1.Pod, host string) error {
+	// TODO: can we distingush between fully bound and waiting for binding to complete?
+	// If it's waiting for binding to complete, then we can return early here instead of
+	// spawning a go routine for bindVolumes to do nothing...
+	areVolumesBound, err := sched.config.VolumeBinder.ArePodVolumesBound(assumed)
```
This needs to check tmpData.allbound instead in case all the pvcs got bound in the middle.

msau42 commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -365,6 +380,12 @@ func (c *configFactory) invalidatePredicatesForPv(pv *v1.PersistentVolume) {
+		invalidPredicates.Insert("NoVolumeBindConflict")
+	}
+
 	c.equivalencePodCache.InvalidateCachedPredicateItemOfAllNodes(invalidPredicates)
```
Question for @kubernetes/sig-scheduling-pr-reviews.  Is the predicate cache per pod per node?  Or just per node?  I was trying to follow this invalidation method and could not find pod references.

bsalamat replied:
EquivalenceCache is a 3 level map structure. The first level key is a node name. The second level key is the predicate key, and the last level key is equivalence hash which is extracted from a pod. So, you may not see a reference to a pod directly. You may only see an equivalence hash.

msau42 commented on pkg/controller/volume/persistentvolume/topology_binder.go:
```diff
@@ -0,0 +1,473 @@
+// scheduling requirements.
+type TopologyAwareVolumeBinder interface {
+
+	// ArePodVolumesBounds returns true if all the Pod's volumes are fully bound
```
Remove this

msau42 commented on pkg/controller/volume/persistentvolume/topology_binder.go:
```diff
@@ -0,0 +1,473 @@
+	BindPodVolumes(assumedPod *v1.Pod) error
+
+	// IsVolumeBindingRequired returns true if API updates are needed for volume binding or provisioning
+	IsVolumeBindingRequired(assumedPod *v1.Pod) bool
```
Remove this. Incorporate into assume and bind functions instead.

msau42 committed:
Refactor PV selection into a common call for scheduler and PV controller

msau42 committed:
Refactor bindVolumeToClaim() into 1) forming new PV object 2) making API and cache update

msau42 committed:
In findMatchingVolumes:
- Delay binding until Pod is scheduled
- Check for PV node affinity and access modes
- Exclude volumes from being matching

msau42 committed:
Add assume cache for PVs

msau42 committed:
Library for scheduler volume binding

msau42 committed:
Scheduler volume cache plumbing and predicate invalidation

msau42 committed:
Add predicate to find volume matches

msau42 committed:
Add assume/bind volume functions to scheduler

msau42 committed:
update build files

msau42 committed:
Add import restriction

msau42 commented:
I've fixed all the TODOs I was planning on doing for this release, and also fixed the commit ordering.  I'll keep on writing more tests.

bsalamat commented on plugin/pkg/scheduler/scheduler.go:
```diff
@@ -241,6 +242,39 @@ func (sched *Scheduler) assume(assumed *v1.Pod, host string) error {
 	// immediately.
 	assumed.Spec.NodeName = host
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.VolumeTopologyBinding) {
```
Could you please place this code in a separate function and call it here?

msau42 replied:
Done

bsalamat commented on plugin/pkg/scheduler/scheduler.go:
```diff
@@ -241,6 +242,39 @@ func (sched *Scheduler) assume(assumed *v1.Pod, host string) error {
+				invalidPredicates := sets.NewString("NoVolumeBindConflict", "NoVolumeZoneConflict")
+				sched.config.Ecache.InvalidateCachedPredicateItemOfAllNodes(invalidPredicates)
+			}
+			return nil
```
I think we should return an error and start a go routine to bind the volumes instead of doing the volume binding in the pod binding function.

msau42 commented on pkg/controller/volume/persistentvolume/topology_binder.go:
```diff
@@ -0,0 +1,482 @@
+	}
+
+	// Mark cache with all the matches for each PVC for this node
+	b.tmpData.nodeBindings[node.Name] = claimsToBind
```
This needs to be protected by a mutex because FindPodVolumes() can be called in parallel across nodes

msau42 commented on plugin/pkg/scheduler/scheduler.go:
```diff
@@ -310,6 +382,11 @@ func (sched *Scheduler) scheduleOne() {
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.VolumeTopologyBinding) {
+		// TODO: find a better way to do this
+		sched.config.VolumeBinder.InitTmpData(pod)
```
@bsalamat this is another place I would greatly appreciate your feedback.

I want to be able to store some temporary data with the pod as it goes through the predicate->priorities->assumeVolume.  The way I have done this here is keep a temporary struct in the volume binder and clear it at the beginning of scheduleOne().

Another way I was considering was to keep my own pod cache with the temporary data.  But then I would also have to add a TTL to prevent memory leaks, which is more heavyweight than the current approach.

I need to store two pieces of information:
* Whether or not all of the pod's PVCs are bound at the beginning of scheduleOne().  This is to handle the case where PVCs can become bound in the middle of running predicates.  This is needed for volume predicate correctness because this PR is changing existing volume predicates to temporarily return true if the pod's PVCs are unbound, but they need to be re-evaluated once the volumes are actually bound.
* When the Pod's PVCs are unbound, cache the returned matching PVs from the VolumeBinding predicate per node.  That way, in the priority function (future), and in the assume volume function, we do not need to search again for the best PV matches.  Every time we search, we need to iterate through all the PVs.

msau42 replied:
I think I can eliminate the first reason.  It was mainly to handle the volume zone predicate under the old design that applied to all PV binding.  But now that this new behavior is restricted, the volume zone predicate is only handling the PVs with the old behavior.  And the new behavior PVs will be handled with the volume bind predicate.  So there's only one predicate that we need to serialize bound state with.

But the 2nd reason for the PV matches cache is still good to have.  I could look into the TTL approach or use a pod informer to delete pods that no longer need scheduling.

msau42 commented:
/test pull-kubernetes-unit

msau42 commented on plugin/pkg/scheduler/scheduler.go:
```diff
@@ -233,6 +234,76 @@ func (sched *Scheduler) preempt(preemptor *v1.Pod, scheduleErr error) (string, e
+		}
+		if bindingRequired {
+			if sched.config.Ecache != nil {
+				invalidPredicates := sets.NewString("NoVolumeBindConflict", "NoVolumeZoneConflict")
```
I think volume zone predicate doesn't need to be invalidated because it's not handling pvcs with the new wait binding mode.

msau42 commented on pkg/controller/volume/persistentvolume/scheduler_binder.go:
```diff
@@ -0,0 +1,511 @@
+}
+
+// FindPodVolumes internally caches the matching PVs per node in tmpData.nodeBindings
+func (b *volumeBinder) FindPodVolumes(pod *v1.Pod, nodeName string) (needsBinding bool, foundMatches bool, err error) {
```
Try splitting this up into 3 separate calls to not overload return values.

msau42 replied:
Or what about these as return values:
* boundVolumesSatisfied bool
* unboundVolumesSatisfied bool 
* err error

msau42 commented on pkg/controller/volume/persistentvolume/scheduler_binder.go:
```diff
@@ -0,0 +1,511 @@
+	// TODO: Need TmpCache for PVC for dynamic provisioning
+	pvcCache  corelisters.PersistentVolumeClaimLister
+	nodeCache corelisters.NodeLister
+	pvCache   PVTmpCache
```
Better name

msau42 commented on pkg/controller/volume/persistentvolume/scheduler_binder.go:
```diff
@@ -0,0 +1,511 @@
+	pvCache   PVTmpCache
+
+	// tmpData is not re-entrant and has to be cleared at the beginning of processing a new Pod
+	tmpData *tmpData
```
Better name. Rework into pod cache

msau42 commented on plugin/pkg/scheduler/scheduler.go:
```diff
@@ -233,6 +234,76 @@ func (sched *Scheduler) preempt(preemptor *v1.Pod, scheduleErr error) (string, e
+				sched.config.Ecache.InvalidateCachedPredicateItemOfAllNodes(invalidPredicates)
+			}
+
+			go func() {
```
Queue up pods and only have one binder routine.

msau42 commented on pkg/controller/volume/persistentvolume/index.go:
```diff
@@ -106,7 +107,13 @@ func (pvIndex *persistentVolumeOrderedIndex) findByClaim(claim *v1.PersistentVol
 }
 
-func findMatchingVolume(claim *v1.PersistentVolumeClaim, volumes []*v1.PersistentVolume) (*v1.PersistentVolume, error) {
+func findMatchingVolume(
```
Add more comments throughout

msau42 committed:
Cache for pod bindings

krzyzacy commented:
/retest

msau42 commented:
This is mostly ready for review.  

Updates:
* got rid of tmpdata, replaced with pod volume binding cache
* rework FindPodVolumes to not overload return values
* renamed TmpCache to AssumeCache

Remaining TODOs:
* Change volume binding to be queued to a single go routine, instead of one go routine per pod.
* Refactor scheduler volume library code to not pull in the entire persistentvolume package.
* remove sc api cherrypick once it is merged

msau42 commented on pkg/controller/volume/persistentvolume/scheduler_binder_cache.go:
```diff
@@ -0,0 +1,135 @@
+	switch t := obj.(type) {
+	case *v1.Pod:
+		pod = t
+	case cache.DeletedFinalStateUnknown:
```
When does this happen?  If it is a common case, then I may still need to have a TTL.

msau42 commented on pkg/controller/volume/persistentvolume/scheduler_binder_cache.go:
```diff
@@ -0,0 +1,135 @@
+	if len(pod.Spec.NodeName) != 0 {
+		return false
+	}
+	return pod.Status.Phase == v1.PodPending
```
@bsalamat I noticed that in the scheduler, it only disallows pod phases "succeeded" and "failed".  When would you have a pod with no node name, but is in "running" or "unknown" state, and you would want the scheduler to schedule it?

msau42 replied:
Discussed with @bsalamat.  I'll hook this cache into the scheduler's pod informer handlers + scheduling error handlers instead of doing my own thing.

anfernee commented on plugin/pkg/scheduler/scheduler.go:
```diff
@@ -230,6 +234,76 @@ func (sched *Scheduler) preempt(preemptor *v1.Pod, scheduleErr error) (string, e
+	// * Retry volume binding if dynamic provisioning fails.
+	// * Bind the Pod to the Node once all volumes are bound.
+	sched.config.Error(assumed, err)
+	sched.config.Recorder.Eventf(assumed, v1.EventTypeWarning, "FailedScheduling", "%v", err)
```
is it a error and warning for the waiting case?

anfernee reviewed (commented):
despite couple of nits, it /lgtm

anfernee commented on plugin/pkg/scheduler/scheduler.go:
```diff
@@ -230,6 +239,97 @@ func (sched *Scheduler) preempt(preemptor *v1.Pod, scheduleErr error) (string, e
+		// * Retry volume binding if dynamic provisioning fails.
+		// * Bind the Pod to the Node once all volumes are bound.
+		sched.config.Error(assumed, err)
+		sched.config.Recorder.Eventf(assumed, v1.EventTypeWarning, "FailedScheduling", "%v", err)
```
Is it a warning for the waiting case?

msau42 replied:
I guess if the reason is "VolumeBindingFailed", then it should be a warning.  Otherwise, it should be normal.  But it's also not a "successful" event.

jsafrane commented on pkg/controller/volume/persistentvolume/index.go:
```diff
@@ -138,9 +164,24 @@ func findMatchingVolume(claim *v1.PersistentVolumeClaim, volumes []*v1.Persisten
+				// is satisfied by the node
+				err := volumeutil.CheckNodeAffinity(volume, node.Labels)
+				if err != nil {
+					return nil, nil
```
Should the scheduler find a better PV in this case (i.e. `continue`)?

msau42 replied:
ah good catch, will also add a test case

jsafrane commented on pkg/controller/volume/persistentvolume/scheduler_assume_cache.go:
```diff
@@ -0,0 +1,283 @@
+	return objInfo
+}
+
+func (c *assumeCache) Get(objName string) interface{} {
```
Should it return `error`? `getObjInfo` would deserve it too.

msau42 commented:
I'm going to defer refactoring the new code to be in a separate package to the next release.  It is causing too much code churn and will make this already big PR even harder to review.

jsafrane commented on pkg/controller/volume/persistentvolume/scheduler_assume_cache.go:
```diff
@@ -0,0 +1,283 @@
+
+	objInfo := c.getObjInfo(name)
+	if objInfo == nil {
+		return fmt.Errorf("Assume: %v %q not found in cache", c.description, name)
```
Errors should start with lowercase (applies to whole file). Unless 'Assume' is name of something and it gets logged from other assume errors too.

msau42 replied:
I was referring to the name of the function, but I can change the errors to not reference function names.

jsafrane commented on pkg/controller/volume/persistentvolume/scheduler_assume_cache.go:
```diff
@@ -0,0 +1,283 @@
+		return fmt.Errorf("Assume: %v %q is out of sync", c.description, name)
+	}
+
+	// Only update the cached PV
```
`cached object`

jsafrane commented on pkg/controller/volume/persistentvolume/scheduler_binder.go:
```diff
@@ -0,0 +1,412 @@
+//    b. Invokes all priority functions.  Future/TBD
+//    c. Selects the best node for the Pod.
+//    d. Cache the node selection for the Pod. (Assume phase)
+//       i.  If PVC binding is required, cache the PV->PVC binding decision for that node.
```
for that node -> for that Pod?

jsafrane commented on pkg/controller/volume/persistentvolume/scheduler_binder.go:
```diff
@@ -0,0 +1,412 @@
+	// FindPodVolumes checks if all of a Pod's PVCs can be satisfied by the node.
+	//
+	// If a PVC is bound, it checks if the PV's NodeAffinity matches the Node.
+	// Otherwise, it tries to find an available PV to bind to the PVC.
```
It would be nice to note that it actually stores the found PV and `AssumePodVolume` uses the stored PVs. From the description it seems `FindPodVolumes` just returns "yes, there are PVs that would satisfy the pod" and some other function actually finds them.

jsafrane reviewed (commented):
Half way through the fourth commit, will continue tomorrow. It looks well thought, I can't find anything serious.

davidz627 commented on plugin/pkg/scheduler/scheduler.go:
```diff
@@ -259,49 +264,70 @@ func (sched *Scheduler) assumeAndBindVolumes(assumed *v1.Pod, host string) error
-	})
+	for {
+		if quit := workFunc(); quit {
+			glog.V(4).Infof("bindVolumesWorker shutting down")
```
should this break from the loop since its shutting down?

davidz627 commented on test/e2e/storage/persistent_volumes-local.go:
```diff
@@ -19,41 +19,48 @@ package storage
+	DirectoryLocalVolumeType localVolumeType = "dir"
 	// creates a tmpfs and mounts it
-	TmpfsLocalVolumeType LocalVolumeType = "tmpfs"
+	TmpfslocalVolumeType localVolumeType = "tmpfs"
```
Tmpfs**L**ocalVolumeType?

msau42 commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -1409,27 +1442,47 @@ type VolumeNodeChecker struct {
 // The requirement is that any pod that uses a PVC that is bound to a PV with topology constraints
 // must be scheduled to a node that satisfies the PV's topology labels.
-func NewVolumeNodePredicate(pvInfo PersistentVolumeInfo, pvcInfo PersistentVolumeClaimInfo, client clientset.Interface) algorithm.FitPredicate {
+func NewVolumeNodePredicate(pvInfo PersistentVolumeInfo, pvcInfo PersistentVolumeClaimInfo) algorithm.FitPredicate {
```
I should just remove this.  This predicate is for an alpha feature so we don't support upgrades anyway.

msau42 commented on plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go:
```diff
@@ -166,11 +166,21 @@ func defaultPredicates() sets.String {
+		),
+
+		// Fit is determined by volume topology requirements.
+		factory.RegisterFitPredicateFactory(
```
See if this can be registered only if feature gate is enabled.  Then we also don't have to add it to the compatibility test.

msau42 commented:
Temporarily removed e2e commit while I resolve major merge conflicts.

msau42 commented on plugin/pkg/scheduler/algorithmprovider/defaults/compatibility_test.go:
```diff
@@ -337,8 +337,7 @@ func TestCompatibility_v1_Scheduler(t *testing.T) {
 			{"name": "GeneralPredicates"},
 			{"name": "TestServiceAffinity", "argument": {"serviceAffinity" : {"labels" : ["region"]}}},
-			{"name": "TestLabelsPresence",  "argument": {"labelsPresence"  : {"labels" : ["foo"], "presence":true}}},
-			{"name": "NoVolumeNodeConflict"}
```
@bsalamat is this ok? I want to remove my alpha feature predicate from the defaults, so it will break backwards compatibility with prior versions that are using the alpha feature.

msau42 replied:
Make sure to put the action required label on this PR.

msau42 commented on plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go:
```diff
@@ -165,15 +165,19 @@ func defaultPredicates() sets.String {
-		// Fit is determined by volume zone requirements.
-		factory.RegisterFitPredicateFactory(
-			"NoVolumeNodeConflict",
+	if utilfeature.DefaultFeatureGate.Enabled(features.VolumeScheduling) {
```
@bsalamat is this an acceptable way to add new predicates under an alpha feature?

msau42 replied:
Ah actually, this is not working.  The predicates list is initialized during init(), which seems to be before the point that the feature gates are set.

jsafrane commented on pkg/controller/volume/persistentvolume/scheduler_binder.go:
```diff
@@ -0,0 +1,414 @@
+	pv *v1.PersistentVolume
+}
+
+func (b bindingInfo) String() string {
```
Unit test coverage indicates that this is never called. Is it used anywhere?

msau42 replied:
I had added it for debugging, but don't need it anymore. Will remove.

msau42 replied:
I remember now. It's used to print out unit test errors.  Will add comment.

jsafrane commented on pkg/controller/volume/persistentvolume/scheduler_binder.go:
```diff
@@ -0,0 +1,414 @@
+// arePodVolumesBound returns true if all volumes are fully bound
+func (b *volumeBinder) arePodVolumesBound(pod *v1.Pod) bool {
+	for _, vol := range pod.Spec.Volumes {
+		if isBound, _, _ := b.isVolumeBound(pod.Namespace, &vol, true); !isBound {
```
Why `true` here? It's called by `AssumePodVolumes` and if there are some PVCs waiting for `annBindCompleted`, then `AssumePodVolumes` will try to update `pvc.spec.volumeName`. But it's already updated...

IMO, `false` would do the same job, saving some cycles in `AssumePodVolumes` and one `BindPodVolumes` call. There is nothing to do, scheduler is just waiting for PV controller to finish the binding. Am I missing something?

msau42 replied:
Right now, `bindVolumesWorker` in the scheduler is the only one that adds the pod back to the scheduler queue, so everything must pass through that currently, even if we're just waiting and there's nothing to do.  I could short circuit it in the `assumeVolumes` as well to save some cycles.  I had left that as a future TODO but I can try to do it now.

jsafrane reviewed (commented):
Finished the review, can't find anything serious, just small issues.

msau42 commented on pkg/controller/volume/persistentvolume/scheduler_binder_cache.go:
```diff
@@ -0,0 +1,135 @@
+	defer c.mutex.Unlock()
+
+	nodeBinding, ok := c.bindings[pod]
+	if !ok {
```
Discussed with @bsalamat.  Will hook this cache into the scheduler's informer handlers and scheduling error handlers so that it will get cleaned up whenever a pod fails scheduling, a pod successfully schedules, or a pod no longer needs to be scheduled.

msau42 commented:
/priority important-soon

msau42 commented:
/kind feature

bsalamat commented on plugin/pkg/scheduler/volumebinder/volume_binder.go:
```diff
@@ -0,0 +1,80 @@
+
+// Run starts a goroutine to handle the binding queue with the given function.
+func (b *VolumeBinder) Run(bindWorkFunc func(), stopCh <-chan struct{}) {
+	if utilfeature.DefaultFeatureGate.Enabled(features.VolumeScheduling) {
```
I guess we don't need to check feature gate here.

bsalamat commented on plugin/pkg/scheduler/volumebinder/volume_binder.go:
```diff
@@ -0,0 +1,80 @@
+
+// DeletePodBindings will delete the cached volume bindings for the given pod.
+func (b *VolumeBinder) DeletePodBindings(pod *v1.Pod) {
+	if utilfeature.DefaultFeatureGate.Enabled(features.VolumeScheduling) {
```
similar to above, probably we don't need the check.

msau42 replied:
DeletePodBindings will grab/release a lock, which shouldn't really see much contention, but who knows?  I guess it's only called outside the main scheduler thread (via informer events or error handler).

bsalamat replied:
The alternative is to put the check on the call site (if there are not many places where this is called). That would make things a bit more clear IMO.

msau42 replied:
I added all the calls in this [commit](https://github.com/kubernetes/kubernetes/pull/55039/commits/47ea0839b65d655b10f7c9ae916801d0458a8be4).   There are 6 callers.

bsalamat reviewed (commented):
@msau42 Do you think you can add a scheduler integration test?

msau42 commented on plugin/pkg/scheduler/scheduler.go:
```diff
@@ -233,6 +239,99 @@ func (sched *Scheduler) preempt(preemptor *v1.Pod, scheduleErr error) (string, e
+			err = fmt.Errorf("Volume binding started, waiting for completion")
+		}
+
+		// Always fail scheduling regardless of binding success.
```
Could you clarify this?  Are you saying that a separate informer event could cause the pod to be added to back to the queue and be processed in scheduleOne() while we are still binding in another thread?

When I look at scheduleOne() and bind() which is in a separate go routine, I see that bind() is updating the Pod outside of the main thread on error.

bsalamat replied:
> Could you clarify this? Are you saying that a separate informer event could cause the pod to be added to back to the queue and be processed in scheduleOne() while we are still binding in another thread?

Yes, it actually does cause this, but my concern is that with this change two threads compete with one another to mark the pod unschedulable. This is a bit of a concern. In case of updating the pod in bind(), scheduler main thread is already done with the pod and may not update it at the same time.

msau42 replied:
Which part of the main thread are you concerned that it is competing with?  In `assumeAndBindVolumes`, we do not update the Pod if we queued it to the volume BindQueue.  And if we queued the pod for volume binding, the `assumeAndBindVolumes()` will return an error so that `scheduleOne()` will return immediately and not continue to `assume`.

bsalamat replied:
You are right, @msau42. I thought scheduler would continue and would update the pod in its main thread, but that's not the case. The only problem here is if the pod is updated in the middle of the volume binding process. In that case it may be retried by the scheduler, but that should be fine.

msau42 replied:
Thanks for the clarification. I'll add some more comments to try to make it more clear.

msau42 commented on plugin/pkg/scheduler/scheduler.go:
```diff
@@ -233,6 +239,99 @@ func (sched *Scheduler) preempt(preemptor *v1.Pod, scheduleErr error) (string, e
+			sched.config.PodConditionUpdater.Update(assumed, &v1.PodCondition{
+				Type:    v1.PodScheduled,
+				Status:  v1.ConditionFalse,
+				Reason:  "SchedulerError",
```
The error returned here is actually a real error, like the object was malformed, or the AssumeVolume failed because the object is out of date.

msau42 commented:
/retest

msau42 commented:
All comments should be addressed.  Still working on an integration test.

msau42 committed:
Add scheduler and pv controller integration test

msau42 commented:
Still working on integration tests, but pushed what I have so far.

msau42 commented on test/integration/scheduler/volume_binding_test.go:
```diff
@@ -0,0 +1,405 @@
+				makePVC("pvc-w-canbind-3", config.ns, &classWait, ""),
+			},
+		},
+		// TODO: this testcase is failing with pod cache modification error
```
I need to debug this failing testcase

msau42 replied:
Found it, needed to make a copy of the assumedPod object before updating it.

bsalamat replied:
Yes! I am glad that the test caught the issue.

bsalamat commented on test/integration/scheduler/volume_binding_test.go:
```diff
@@ -0,0 +1,405 @@
+		}
+
+		// TODO: Verify PVC/PV binding
+		// TODO: Verify events on Pods and PVCs
```
To ensure your test cases are independent and don't conflict, please delete the pod (and probably PVs and PVCs) at the end of each case.

msau42 replied:
done

bsalamat commented on test/integration/scheduler/volume_binding_test.go:
```diff
@@ -0,0 +1,405 @@
+	pv := &v1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        name,
+			Annotations: map[string]string{},
```
nit: I guess you can drop this line.

msau42 replied:
this is needed to set the PV node affinity below

msau42 commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -365,6 +387,12 @@ func (c *configFactory) invalidatePredicatesForPv(pv *v1.PersistentVolume) {
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.VolumeScheduling) {
+		// Add/delete impacts the available PVs to choose from
+		invalidPredicates.Insert(predicates.CheckVolumeBinding)
```
Also need to invalidate predicate on PVC update when volumeName changes.

msau42 replied:
done

msau42 commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -223,6 +223,13 @@ func NewConfigFactory(
+						return
+					}
+					// Volume binder only wants to keep unassigned pods
+					c.volumeBinder.DeletePodBindings(pod)
```
@bsalamat can you verify that these are all the right places to clean up my cache?   I want to clean it up anytime the Pod is no longer unassigned.

And also the scenario where the pod gets deleted/updated while being processed in scheduleOne().

bsalamat replied:
This looks correct to me. This delete will be triggered when:
1. An unassigned pod is scheduled and bound.
2. An unassigned pod is actually deleted.

I am just not so sure about a case where an unassigned pod is updated and its volume claims change. Do we allow such an update at all? If we do, I guess we need invalidating the cache.

msau42 replied:
Thanks.  You cannot modify pod.Spec.Volumes.

bsalamat commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -211,6 +223,13 @@ func NewConfigFactory(
 					if err := c.podQueue.Delete(obj.(*v1.Pod)); err != nil {
 						runtime.HandleError(fmt.Errorf("unable to dequeue %T: %v", obj, err))
 					}
+					pod, ok := obj.(*v1.Pod)
```
please move this line to the beginning of this function and don't add "ok" and don't check it. We prefer to see a panic if the obj is not a pod. See `UpdateFunc` as an example.

msau42 replied:
done

bsalamat commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -436,6 +454,21 @@ func (c *configFactory) invalidatePredicatesForPvc(pvc *v1.PersistentVolumeClaim
 	}
 }
 
+func (c *configFactory) invalidatePredicatesForPvcUpdate(old, new *v1.PersistentVolumeClaim) {
```
Ideally, we would like to have integration tests for these changes, but given that this PR is already too large and we are close to the code freeze, you may want to add the tests in a separate PR.

msau42 replied:
Will open a bug to add integration tests for predicate invalidation.

msau42 commented on plugin/pkg/scheduler/scheduler.go:
```diff
@@ -252,6 +252,7 @@ func (sched *Scheduler) assumeAndBindVolumes(assumed *v1.Pod, host string) error
 	if utilfeature.DefaultFeatureGate.Enabled(features.VolumeScheduling) {
 		allBound, bindingRequired, err := sched.config.VolumeBinder.Binder.AssumePodVolumes(assumed, host)
 		if err != nil {
+			assumed = assumed.DeepCopy()
```
@bsalamat instead of doing the DeepCopy() here, I'm thinking it might be better to do it outside in scheduleOne() here: 
https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/scheduler.go#L325

WDYT?

bsalamat replied:
Makes sense.

msau42 replied:
The sequence of events to trigger this issue was:
1. Pod fails predicates, pod condition was updated with the predicate error.
2. Pod passes predicates
3. Pod fails bind, pod condition was updated with the bind error.  It seems like the Conditions in `assumedPod` is the same ptr as in `pod` because `assumedPod` is not a DeepCopy of `pod`.

msau42 replied:
Updated, let me know when I can squash

bsalamat replied:
Thanks! Looks good. Please go ahead and squash.

bsalamat replied:
You actually fixed a bug that existed in the scheduler code for a while. The scenario that you described could happen without your changes (and in fact, we had seen some api cache panics which I think was caused by this issue).

msau42 commented:
/retest

msau42 commented:
/retest

msau42 committed:
Local e2e tests

msau42 commented:
/retest

msau42 commented:
/retest

jsafrane commented:
/approve
& lgtm the pvcontroller changes

bsalamat commented:
/lgtm
for scheduler changes.

msau42 commented:
/assign @liggitt 

Would you be able to approve the kubectl and auth changes?

liggitt commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go:
```diff
@@ -388,7 +388,9 @@ func ClusterRoles() []rbac.ClusterRole {
 				rbac.NewRule(Read...).Groups(policyGroup).Resources("poddisruptionbudgets").RuleOrDie(),
-				rbac.NewRule(Read...).Groups(legacyGroup).Resources("persistentvolumeclaims", "persistentvolumes").RuleOrDie(),
+				rbac.NewRule(Read...).Groups(legacyGroup).Resources("persistentvolumeclaims").RuleOrDie(),
+				rbac.NewRule("update", "get", "list", "watch").Groups(legacyGroup).Resources("persistentvolumes").RuleOrDie(),
```
what in the PV spec is the scheduler updating?

msau42 replied:
It's updating pv.Spec.ClaimRef, and the annBoundByController annotation.

msau42 commented:
/test pull-kubernetes-kubemark-e2e-gce

msau42 commented:
Added new [commit](https://github.com/kubernetes/kubernetes/pull/55039/commits/3db168a6864d0c6ebd48bbf46f475eb4a504afcd) to:
* Enable new scheduler cluster rolebindings under feature gate
* Only setup storageclass informer under feature gate
* Only create volumebinder under feature gate
* Add predicate errors to the `nodesWherePreemptionMightHelp` list

@bsalamat @liggitt

msau42 commented:
/retest

msau42 committed:
Add note to feature gate

msau42 commented:
/retest

liggitt commented on pkg/controller/volume/persistentvolume/pv_controller.go:
```diff
@@ -272,15 +302,19 @@ func (ctrl *PersistentVolumeController) syncUnboundClaim(claim *v1.PersistentVol
 			// No PV could be found
 			// OBSERVATION: pvc is "Pending", will retry
-			if v1helper.GetPersistentVolumeClaimClass(claim) != "" {
+			if v1helper.GetPersistentVolumeClaimClass(claim) != "" && !delayBinding {
```
reordering avoids the negative case:
```go
switch {
case delayBinding:
...
case v1helper.GetPersistentVolumeClaimClass(claim) != "":
...
default:
...
}
```

msau42 replied:
done.  @jsafrane can you also take a look at this update? i squashed it into the 3rd commit.

jsafrane replied:
checked pv_controller.go,  lgtm

liggitt commented:
needs some squashing to avoid churn in commits

I still think it would be better to maintain the division of responsibilities between the pod scheduler and a volume scheduler. That keeps the data flows between components clearer and allows custom pod schedulers to avoid needing to take responsibility for binding local PVs

That said, to unblock continued work on this feature, I'll provisionally approve the gated change to the scheduler role since it is not enabled by default, and there is an issue for reworking this before bringing the feature to beta. I do not expect update PV permission to become part of the scheduler role by default.

liggitt commented on pkg/controller/volume/persistentvolume/pv_controller.go:
```diff
@@ -37,6 +38,7 @@ import (
 	"k8s.io/kubernetes/pkg/controller/volume/events"
+	"k8s.io/kubernetes/pkg/features"
 	"k8s.io/kubernetes/pkg/util/goroutinemap"
 	"k8s.io/kubernetes/pkg/util/goroutinemap/exponentialbackoff"
```
weird... double import?

liggitt replied:
nm, misread beacuse of weird line wrapping

msau42 commented:
Squashed the last commit into 'scheduler plumbing', 'predicate', and 'build files' commits.

Thanks for your feedback @liggitt.  We'll definitely see what we can do to improve the design.

bsalamat commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -468,8 +539,15 @@ func (c *configFactory) addPodToCache(obj interface{}) {
 	// NOTE: Updating equivalence cache of addPodToCache has been
 	// handled optimistically in InvalidateCachedPredicateItemForPodAdd.
+
+	if c.volumeBinder != nil {
```
This is not needed. You have code to delete the bindings in the delete event handler for unassigned pods.

msau42 commented:
/assign @thockin for pkg/features and pkg/kubectl changes

bsalamat commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -490,6 +568,11 @@ func (c *configFactory) updatePodInCache(oldObj, newObj interface{}) {
 	c.invalidateCachedPredicatesOnUpdatePod(newPod, oldPod)
 	c.podQueue.AssignedPodUpdated(newPod)
+
+	if c.volumeBinder != nil {
```
Similar to above. Why do we need to repeat this code?

bsalamat commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -535,6 +618,11 @@ func (c *configFactory) deletePodFromCache(obj interface{}) {
 	c.invalidateCachedPredicatesOnDeletePod(pod)
 	c.podQueue.MoveAllToActiveQueue()
+
+	if c.volumeBinder != nil {
```
and here too

bsalamat commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -1066,12 +1159,22 @@ func (factory *configFactory) MakeDefaultErrorFunc(backoff *util.PodBackoff, pod
 				if err == nil {
 					if len(pod.Spec.NodeName) == 0 {
 						podQueue.AddUnschedulableIfNotPresent(pod)
+					} else {
```
Please remove this one too.

msau42 replied:
What about the scenario where the pod is updated/deleted while it is being processed in `scheduleOne`?  The two checks here were supposed to be for that case.

bsalamat replied:
The pod is unassinged while it is being processed in `scheduleOne`. So, if it is updated or deleted, we will get the corresponding event for the unassigned pod.

msau42 replied:
What if:
1. Start processing pod in scheduleOne
2. Pod gets deleted, unassigned event comes and clears cache
3. The processing in scheduleOne updates the cache with the pod
4. Assume/bind fails because pod doesn't exist anymore
5. Cache needs to be cleaned up somehow

bsalamat replied:
The scenario is valid, but in that case pod will not be assigned (len(pod.Spec.NodeName) == 0) and the pod will not be found, so err != nil. So, the check you have added below to `if errors.IsNotFound(err) {` should suffice.

msau42 replied:
What if it was an update event that caused the pod to be assigned?

bsalamat replied:
The way our event handlers are set up is that they have a filter that check for assignment of pods. So, whenever a pod is assigned, we received one "delete" event for the unassigned pod followed by an "add" event for the assigned pod. In other words, an unassigned pod won't become assigned in an update.

bsalamat replied:
Had an offline discussion with @msau42. This is correct.

bsalamat commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -1066,12 +1159,22 @@ func (factory *configFactory) MakeDefaultErrorFunc(backoff *util.PodBackoff, pod
 					glog.Warningf("A pod %v no longer exists", podID)
-					return
+
+					if factory.volumeBinder != nil {
```
And this one.

thockin commented:
Sorry I missed the conversation earlier, @liggitt .  I 100% hear the concerns, and we're very open to clever ideas.  We didn't find any good options yet.

/approve

bsalamat commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -1066,12 +1143,22 @@ func (factory *configFactory) MakeDefaultErrorFunc(backoff *util.PodBackoff, pod
+					if factory.volumeBinder != nil {
+						// Volume binder only wants to keep unassigned pods
+						factory.volumeBinder.DeletePodBindings(origPod)
+						return
```
please move `return` out of `if factory.volumeBinder != nil {`.

bsalamat commented:
/lgtm
for scheduler changes
